src/msw/joystick.cpp

```diff
@@ -37,8 +37,127 @@
 
 #include <regstr.h>
 
+// Use optimised count trailing zeros where available.
+#ifdef __GNUC__
+    #define ctz(x) __builtin_ctz(x)
+#else
+int ctz(unsigned x) {
+   int n;
+
+   if (x == 0) return(32);
+   n = 1;
+   if ((x & 0x0000FFFF) == 0) {n = n +16; x = x >>16;}
+   if ((x & 0x000000FF) == 0) {n = n + 8; x = x >> 8;}
+   if ((x & 0x0000000F) == 0) {n = n + 4; x = x >> 4;}
+   if ((x & 0x00000003) == 0) {n = n + 2; x = x >> 2;}
+   return n - (x & 1);
+}
+#endif
+
+enum {
+    wxJS_AXIS_X = 0,
+    wxJS_AXIS_Y,
+    wxJS_AXIS_Z,
+    wxJS_AXIS_RUDDER,
+    wxJS_AXIS_U,
+    wxJS_AXIS_V,
+
+    wxJS_AXIS_MAX = 32767,
+    wxJS_AXIS_MIN = -32767,
+    wxJS_MAX_AXES = 6,      // WinMM supports up to 6 axes.
+    wxJS_MAX_BUTTONS = 32,  // WinMM supports up to 32 buttons.
+};
+
 wxIMPLEMENT_DYNAMIC_CLASS(wxJoystick, wxObject);
 
+////////////////////////////////////////////////////////////////////////////
+// Background thread for reading the joystick device
+////////////////////////////////////////////////////////////////////////////
+
+class wxJoystickThread : public wxThread
+{
+public:
+    wxJoystickThread(int joystick);
+    void* Entry() wxOVERRIDE;
+
+private:
+    void      SendEvent(wxEventType type, long ts, int change = 0);
+    int       m_joystick;
+    int       m_buttons;
+    wxWindow* m_catchwin;
+    int       m_polling;
+    JOYINFO   m_joyInfo;
+    JOYINFO   m_lastJoyInfo;
+
+    friend class wxJoystick;
+};
+
+
+wxJoystickThread::wxJoystickThread(int joystick)
+    : m_joystick(joystick),
+      m_buttons(0),
+      m_catchwin(NULL),
+      m_polling(0),
+      m_joyInfo(),
+      m_lastJoyInfo()
+{
+}
+
+void wxJoystickThread::SendEvent(wxEventType type, long ts, int change)
+{
+    wxJoystickEvent jwx_event(type, m_buttons, m_joystick, change);
+
+    jwx_event.SetTimestamp(ts);
+    jwx_event.SetPosition(wxPoint(m_joyInfo.wXpos, m_joyInfo.wYpos));
+    jwx_event.SetZPosition(m_joyInfo.wZpos);
+    jwx_event.SetEventObject(m_catchwin);
+
+    if (m_catchwin)
+        m_catchwin->GetEventHandler()->AddPendingEvent(jwx_event);
```

Comment:
Using `ProcessThreadEvent()` instead of `AddPendingEvent()` would be more clear here.

Reply:
Done.